pkg/frontend/querymiddleware/durations.go

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+// newDurationsMiddleware returns a MetricsQueryMiddleware that evaluates and
+// rewrites duration expressions in the query.
+// As long as durations can be calculated solely from consts and the step
+// this works. When that is not the case anymore, we'll have to implement
+// checks before sharding/splitting/etc middlewares individually to check what
+// we _can_ support.
+func newDurationsMiddleware(
+	logger log.Logger,
+) MetricsQueryMiddleware {
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &durationsMiddleware{
+			next:   next,
+			logger: logger,
+		}
+	})
+}
+
+type durationsMiddleware struct {
+	next   MetricsQueryHandler
+	logger log.Logger
+}
+
+func (d *durationsMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	// Log the instant query and its timestamp in every error log, so that we have more information for debugging failures.
+	logger := log.With(d.logger, "query", req.GetQuery(), "query_timestamp", req.GetStart())
+
+	spanLog, ctx := spanlogger.New(ctx, logger, tracer, "durationsMiddleware.Do")
+	defer spanLog.Finish()
+
+	expr, err := parser.ParseExpr(req.GetQuery())
+	if err != nil {
+		level.Warn(spanLog).Log("msg", "failed to parse query", "err", err)
+		// This middleware focuses on duration expressions, so if the query is
+		// not valid, we just fall through to the next handler.
+		return d.next.Do(ctx, req)
+	}
+
+	expr, err = promql.PreprocessExpr(expr, time.UnixMilli(req.GetStart()), time.UnixMilli(req.GetEnd()), time.Duration(req.GetStep())*time.Second)
+	if err != nil {
+		level.Warn(spanLog).Log("msg", "failed to evaluate duration expressions in query", "err", err)
+		return nil, apierror.New(apierror.TypeBadData, DecorateWithParamName(err, "query").Error())
+	}
+
+	checkVisitor := &durationVisitor{}
+	if err := parser.Walk(checkVisitor, expr, nil); err != nil {
+		level.Warn(spanLog).Log("msg", "the query contains unsupported duration expressions", "err", err)
+		return nil, apierror.New(apierror.TypeBadData, DecorateWithParamName(err, "query").Error())
+	}
+
+	req, err = req.WithExpr(expr)
+	if err != nil {
+		level.Warn(spanLog).Log("msg", "failed to update the query expression after durations expressions were evaluated", "err", err)
+		return nil, apierror.New(apierror.TypeInternal, DecorateWithParamName(err, "query").Error())
+	}
+
+	return d.next.Do(ctx, req)
+}
+
+type durationVisitor struct{}
+
+// Verify that duration expressions only contain durations that Mimir supports.
```

Comment:
```suggestion
// Visit verifies that duration expressions only contain durations that Mimir supports.
```
I just run the linter in my head since there is already a comment 🙈